File: leetcode/make-the-string-great/answer2.java
class Solution {
    public String makeGood(String s) {
        ArrayDeque<Character> stack = new ArrayDeque<>();
        for(char c : s.toCharArray()) {
            if(!stack.isEmpty()) {
                if((stack.peek() ^ c) == 32) {
                    stack.pop();
                    continue;
                }
            }

            stack.push(c);
        }

        StringBuilder sb = new StringBuilder();
        while(!stack.isEmpty()) {
            sb.append(stack.pop());
        }

        return sb.reverse().toString();
    }
}

File: leetcode/valid-parentheses/answer2.kt
class Solution {
    fun isValid(s: String): Boolean {
        val stack: ArrayDeque<Char> = ArrayDeque()

        val table = mapOf(
            ')' to '(',
            ']' to '[',
            '}' to '{'
        )

        for(ch in s) {
            if (!table.contains(ch)) {
                stack.addLast(ch)
            } else if(stack.isEmpty() || table[ch] != stack.removeLast()) {
                return false
            }
        }

        return stack.isEmpty()
    }
}